### PR TITLE
Avoid symbolic size-one metadata guards

### LINKED ISSUE #143463: unreasonable ConstraintViolationError when using torch dynamo to compile torch model
### 🐛 Describe the bug

I'm using torch dynamo backend to compile model to export to tensorrt.
```python
    inputs = [torch.randn(1, 3, 28, 288, 512).cuda().to(torch.float16)]
    
    dynamic_h = torch.export.Dim("dim_3", min=224, max=640)
    dynamic_w = torch.export.Dim("dim_4", min=224, max=640)
    dynamic_t = torch.export.Dim("dim_2", min=1, max=200)
    dynamic_shapes={"x": {2:dynamic_t, 3: dynamic_h, 4: dynamic_w}}

    exp_program = torch.export.export(encoder_model, args=tuple(inputs), dynamic_shapes=dynamic_shapes, strict=True)

    trt_model = torch_tensorrt.dynamo.compile(
        exported_program=exp_program,
        assume_dynamic_shape_support=True,
        inputs=inputs, 
        make_refitable=True, 
        disable_tf32=True, 
        debug=True, 
        enabled_precisions={torch.half, torch.float},
        torch_executed_ops={},
        min_block_size=17,
        truncate_double=True,
        use_python_runtime=False)
```
I've confirmed that the h, w dim can be dynamic without constraint error but the dim_2 cannot. It's quite strange.
I trace the cause and found it's in here
```
class GroupNormSpatial(nn.Module):
    """GroupNorm with spatial dimensions ignored."""
    def __init__(self, num_groups, num_channels, epsilon: float = 1e-5, affine=True):
        super().__init__()
        # affine=False # TODO: for tensorrt only
        self.norm_fn = nn.GroupNorm(num_groups=num_groups, num_channels=num_channels, eps=epsilon, affine=affine)

    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
        if int(inputs.ndim) == 5: # video
            b, c, t, h, w = inputs.shape
            inputs = inputs.permute(0,2,1,3,4).flatten(start_dim=0, end_dim=1) # ERROR
            out = self.norm_fn(inputs)
            out = out.reshape(b, t, c, h, w).permute(0,2,1,3,4) # ERROR
            return out
        else: # Image, b c h w -> b c h w
            out = self.norm_fn(inputs)
            return out
```

```
I1218 02:40:31.595000 155463 torch/_utils_internal.py:116] [0/0] CompilationMetrics(compile_id='0/0', frame_key='1', co_name='forward', co_filename='xxx,py', co_firstlineno=50, cache_size=0, accumulated_cache_size=0, guard_count=None, shape_env_guard_count=None, graph_op_count=None, graph_node_count=None, graph_input_count=None, start_time=1734489622.2484767, entire_frame_compile_time_s=None, backend_compile_time_s=None, inductor_compile_time_s=None, code_gen_time_s=None, fail_type="<class 'torch.fx.experimental.symbolic_shapes.ConstraintViolationError'>", fail_reason='Constraints violated (dim_2)! For more information, run with TORCH_LOGS="+dynamic".\n  - Not all values of dim_2 = L[\'x\'].size()[2] in the specified range dim_2 <= 200 satisfy the generated guard Ne(Mod(1, ((L[\'x\'].size()[2] - 1)//2) + 1), 0).\n  - Not all values of dim_2 = L[\'x\'].size()[2] in the specified range dim_2 <= 200 satisfy the generated guard Ne(Mod(1, ((L[\'x\'].size()[2] - 1)//4) + 1), 0).\n  - Not all values of dim_2 = L[\'x\'].size()[2] in the specified range dim_2 <= 200 satisfy the generated guard 9 <= L[\'x\'].size()[2] and L[\'x\'].size()[2] <= 200', fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=set(), compliant_custom_ops=set(), restart_reasons=set(), dynamo_time_before_restart_s=9.347490787506104, has_guarded_code=False, possibly_missed_reinplacing_opportunities=None)
```

### Versions

I'm using ngc torch `nvcr.io/nvidia/pytorch:24.10-py3`

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186032

Root cause: export recorded guards from metadata-only layout, broadcast, and view decisions that used example-shape truth for symbolic size-one checks. In the issue repro, temporal slicing feeds GroupNormSpatial through flatten, reshape, and group_norm; those checks excluded valid values from the declared Dim range.

Use static proofs for optional metadata fast paths that can safely fall back, while preserving guard-based behavior for normal torch.compile broadcasting. Preserve the original requested fake-view shape across retries, avoid specializing unknown source dimensions when reshaping to a size-one target, and directly flatten the statically size-one leading-dimension case needed by the repro. Symbolic group_norm inputs clone into a statically known-safe memory format instead of invoking guard-producing contiguity checks.

After rebasing over #186414, native_group_norm owns concrete input and affine contiguity. Keep only the high-level symbolic input clone here and pass weight and bias through unchanged, avoiding duplicate validation and conversions while retaining the guard-free symbolic path.

Accuracy runs also reapply the existing determinism setup after benchmark model construction. Some model constructors reset global determinism flags, which otherwise makes the eager-vs-eager check fail before the compiler path runs.

Fixes #143463

Generated by my agent

Test Plan:

```bash
.venv/bin/python -m pip install -e . -v --no-build-isolation
.venv/bin/python test/export/test_export.py TestDynamismExpression.test_reshape_view_static_size_one_before_dynamic_dim TestDynamismExpression.test_group_norm_spatial_dynamic_temporal_dim TestDynamismExpression.test_reshape_view_backed_size_oblivious TestDynamismExpression.test_view_unify_cross_symbols -q
.venv/bin/python test/export/test_export.py TestDynamismExpression.test_group_norm_spatial_dynamic_temporal_dim_cuda -q
.venv/bin/python test/export/test_export.py TestExport.test_dim_dynamic -q
.venv/bin/python test/test_dynamic_shapes.py TestUbackedOps.test_unbacked_reshape1 TestUbackedOps.test_backed_size_oblivious_broadcast -q
.venv/bin/python -m py_compile torch/_refs/__init__.py test/export/test_export.py
CPATH=/usr/lib/gcc/x86_64-redhat-linux/11/include lintrunner -a
git diff --check
git diff --cached --check
```

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says @claude review these changes

Updated the PR to address the dynamic-shape CI regressions without the earlier broad symbolic metadata changes. The new diff keeps TensorImpl symbolic layout predicates unchanged, narrows group_norm to make an unconditional contiguous clone only when symbolic layout contiguity is not statically proven, and fixes reshape/collapse metadata to avoid guard-producing size-one decisions while preserving correct strides.

Validation:
- ninja -C build torch_cpu
- python test/export/test_export.py -k test_group_norm_spatial_dynamic_temporal_dim
- python test/export/test_export.py TestDynamismExpression.test_view_unify_cross_symbols
- collapse-view symbolic possibly-size-one stride regression check
- git diff --check
- fresh agent review found no issues

Known local blockers:
- lintrunner -a still fails only on unrelated torch/nn/modules/loss.py pyrefly unexpected-keyword linear_bias
- inductor test invocation is blocked locally by PyTorch CUDA 12.8 / torchvision CUDA 13.0 mismatch

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the failing dynamic-shape CI regressions. The fix keeps the GroupNorm symbolic-size guard removal, narrows Python metadata changes to static-proof fast paths, adds a generic symbolic reshape divisibility regression, and validates the earlier derived-dim and ubacked reshape failures locally.

**Comment by jansel** [comment]
My agent says I updated the PR to avoid symbolic size-one metadata guards in the GroupNormSpatial export repro. Targeted tests now pass locally: TestDynamismExpression.test_group_norm_spatial_dynamic_temporal_dim, TestExport.test_dim_dynamic, TestUbackedOps.test_unbacked_reshape1, plus lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to fix the CI regression from the prior broad flatten fast path. The flatten special case is now limited to the statically size-one leading-dimension pattern needed by the GroupNormSpatial export repro, and the reproduced CpuTests.test_cat_cpu failure passes locally with the torchvision import shim.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the latest CI fallout. The broadcast metadata change is now export-only so normal torch.compile keeps hinted guard_or_false broadcasting behavior, and the flatten as_strided fast path is limited to symbolic-size tensors so concrete/subclass paths such as DTensor cartesian_prod keep the old reshape behavior. Local targeted checks now pass for the original GroupNormSpatial export repro, export max, unbacked reshape, backed-size-oblivious broadcast, memory_format_operators, cholesky_solve_batched_broadcasting, and DTensor cartesian_prod.

**Comment by jansel** [comment]
My agent says I fixed the latest CI regressions by restoring normal torch.compile broadcast guard behavior outside export, keeping export's static-proof broadcast decisions, and restoring the hinted SymbolicShapeMeta layout fast path to avoid proxy/AOT symbolic-layout timeouts. Local targeted coverage now passes for the memory-format Dynamo failure, the proxy conv_transpose3d case, the AOT linalg_lu/tensorinv cases, and the original export regressions.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the latest review feedback by removing the stray no-op SymbolicShapeMeta hunk. A fresh review also caught that symbolic CUDA group_norm still reached guard-producing contiguous checks, so I extended the symbolic group_norm path to clone directly into contiguous format for non-CPU devices and added CUDA coverage for the GroupNormSpatial export regression. Local validation passed for the CPU and CUDA GroupNormSpatial export tests, torch_cpu rebuild, diff checks, py_compile, lintrunner -a, and a fresh review pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR for the latest inductor TorchBench CI failure. The failure was an eager determinism check that happened before Inductor ran: TorchBench torchvision model construction reset cuDNN determinism flags after the benchmark harness configured accuracy-run determinism. The update reapplies the existing determinism setup after model load for accuracy runs, and restores the real compiled graph-break baselines for mnasnet1_0 and mobilenet_v2. Local targeted TorchBench accuracy and graph-break checks now pass for both models.

**Comment by jansel** [comment]
My agent says the TorchBench CSV changes are from reapplying the existing determinism setup after model construction, not from the symbolic-shape metadata fix. The affected torchvision constructors reset cuDNN determinism flags, so reapplying determinism makes the eager-vs-eager accuracy check exercise the intended deterministic compiled path; the higher graph-break counts for mnasnet1_0 and mobilenet_v2 are the real compiled baselines once that masking is removed. I added this context to the PR description, and the latest Dr. CI classifies remaining inductor failures as flaky or broken trunk rather than new regressions.

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -3991,6 +4020,15 @@ def _reshape_view_helper_core_alg(
             idx = idx + 1
             continue
 
+        # A target dimension of length 1 can always be produced by splitting
+        # the current source dimension with an outer length of 1. Do this before
+        # equality checks so symbolic source sizes do not specialize on whether
+        # they are 1.
+        if isinstance(length, utils.IntWithoutSymInt) and length == 1:
```
Comment (bobrenjc93) [inline torch/_refs/__init__.py:4152]: This new size-1 block splits unconditionally, but `prims.split_dim(a_, idx, 1)` inserts a *new* size-1 outer dim and leaves the original source dim at idx+1 — it does not consume it. When the source dim at idx is itself statically 1 and is followed by a symbolic dim, this misaligns the walk. Trace `reshape([1, t, C, H, W]) -> [1, t, C, H*W]` (batch=1 concrete, t dynamic): target[0]=1 splits source dim0(size 1) into (1,1) giving [1,1,t,C,H,W]; then target[1]=t is compared against a_.shape[1]==1, evaluating `t == 1` which installs a guard (t != 1) plus a `1 % t` guard. The old code matched the leading concrete 1 to source dim0 and t to t with no guards. So this unconditionally regresses a very common batch-1 dynamic-shape reshape, directly reintroducing the size-one specialization this PR is trying to remove (the added test happens to have source dim0 = t, not 1, so it doesn't cover this). Suggest only splitting when the source dim isn't statically 1, e.g. `... and length == 1 and not guard_or_false(a_.shape[idx] == 1)`, so statically-1 source dims skip via the existing equality path; and add a test with a leading concrete-1 batch dim.

Comment (jansel) [reply torch/_refs/__init__.py:4152]: My agent says thanks, fixed in the latest update. The reshape helper now only uses the target-size-1 split when the source dimension is not statically known to be 1, using statically_known_true rather than a guard-producing check. I also added test_reshape_view_static_size_one_before_dynamic_dim for reshape([1, t, C, H, W]) -> [1, t, C, H*W].

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and adapted the GroupNorm path to #186414: concrete inputs and affine contiguity remain owned by native_group_norm, while the high-level path keeps only the symbolic input clone required to avoid size-one guards. The supported editable CPU/CUDA build, focused CPU/CUDA export regressions, adjacent dynamic-shape tests, py_compile, and lintrunner -a all pass. CI has restarted on be2c0474df52.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
